storybookjs/storybook
# fix(storybook): Use parameters from first story in group

Issue: #15772

## What I did

Currently, when `transformStoriesRawToStoriesHash` creates a `Group`, the group ends up with the `parameters` of its last child story, because the `parameters` are replaced by each `merge`. This PR changes the merge behavior so that the group will have the parameters of its first child story instead (or, more precisely, the first child story where `parameters` is defined).

This fixes the behavior described in #15772, since `selectStory()` [uses the `parameters.viewMode` from the group](https://github.com/storybookjs/storybook/blob/e6022f09d9f7c4e39726220f2ef4d56b164b6434/lib/api/src/modules/stories.ts#L288-L294) when given the id of a group, even though it's effectively navigating to the group's first child.

Alternatively, the issue could be fixed in the `selectStory()` logic (by having it use `s.children[0].parameters.viewMode` when navigating to `s.children[0]`), but I thought it might be preferable to fix the discrepancy at the story hash level in case the `parameters` field of story groups is used elsewhere.

## How to test

I've updated the stories API unit tests to cover this scenario.

If you want to experience the behavior this was intended to fix, try changing `examples/official-storybook/stories/addon-docs/subcomponents.stories.js` to this:

<details>
  <summary>Stories with different parameters</summary>

  ```ts
  import React from 'react';
  import { DocgenButton } from '../../components/DocgenButton';
  import { ButtonGroup } from '../../components/ButtonGroup';

  export default {
    title: 'Addons/Docs/ButtonGroup',
    component: ButtonGroup,
    subcomponents: { DocgenButton },
  };

  export const DocsMode = () => (
    <ButtonGroup background="#eee">
      <DocgenButton label="This should be shown in docs mode" />
      <DocgenButton label="bar" />
    </ButtonGroup>
  );

  DocsMode.parameters = { viewMode: 'docs' };

  export const StoryMode = () => (
    <ButtonGroup background="#eee">
      <DocgenButton label="This should be shown in story mode" />
      <DocgenButton label="bar" />
    </ButtonGroup>
  );

  StoryMode.parameters = { viewMode: 'story' };
  ```
</details>

Then run `yarn storybook` and navigate to `http://localhost:9011/?path=/docs/addons-docs-buttongroup`. On the `next` branch, the first story is shown but the `viewMode` from the second story is used. On this PR branch, the first story is shown and its `viewMode` is used, as expected.
